File: solutions-by-language/python/0679_24_game.py
# Problem: 679 - 24 Game
 # Description: Given four numbers, determine if they can be combined using +, -, *, / and parentheses to make exactly 24.
 # Date: 2025-08-18

from typing import List

class Solution:
    def judgePoint24(self, cards: List[int]) -> bool:
        nums = [float(x) for x in cards]
        EPS = 1e-6

        def dfs(arr: List[float]) -> bool:
            if len(arr) == 1:
                return abs(arr[0] - 24.0) < EPS

            n = len(arr)
            for i in range(n):
                for j in range(i + 1, n):
                    a, b = arr[i], arr[j]

                    results = [a + b, a - b, b - a, a * b]
                    if abs(b) > EPS:
                        results.append(a / b)
                    if abs(a) > EPS:
                        results.append(b / a)

                    rest = [arr[k] for k in range(n) if k != i and k != j]
                    for val in results:
                        if dfs(rest + [val]):
                            return True
            return False

        return dfs(nums)

File: solutions-by-language/java/0679_24_game.java
// Problem: 679 - 24 Game
// Description: Given four numbers, determine if they can be combined using +, -, *, / and parentheses to make exactly 24.
// Date: 2025-08-18

import java.util.*;

class Solution {
    public boolean judgePoint24(int[] cards) {
        List<Double> nums = new ArrayList<>();
        for (int x : cards) nums.add((double) x);
        final double EPS = 1e-6;
        return dfs(nums, EPS);
    }

    private boolean dfs(List<Double> nums, double EPS) {
        if (nums.size() == 1) {
            return Math.abs(nums.get(0) - 24.0) < EPS;
        }

        int n = nums.size();
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                double a = nums.get(i), b = nums.get(j);
                List<Double> candidates = new ArrayList<>(Arrays.asList(
                    a + b, a - b, b - a, a * b
                ));
                if (Math.abs(b) > EPS) candidates.add(a / b);
                if (Math.abs(a) > EPS) candidates.add(b / a);

                // Remaining numbers after removing i and j
                List<Double> rest = new ArrayList<>();
                for (int k = 0; k < n; ++k) {
                    if (k != i && k != j) rest.add(nums.get(k));
                }

                for (double val : candidates) {
                    rest.add(val);
                    if (dfs(rest, EPS)) return true;
                    rest.remove(rest.size() - 1);
                }
            }
        }
        return false;
    }
}